### PR TITLE
feat: cmake option to skip building mock libraries

### PR DESCRIPTION
**Pull request by dbolduc**
Motivated by #13621 

Consumers of `google-cloud-cpp` who want to test their code can use this flag to install mocks, without needing to build all of our unit tests, internal testing infrastructure.

# Background

We create mocking libraries (which are all header only) and install them by default (even when `BUILD_TESTING=OFF`). For many months, we only did this in Pub/Sub and GCS+gRPC, but in the latest release, we started installing mocks for all GAPICs. (And in this release, for all libraries except compute).

These libraries link to GoogleTest, which can be hard to find. This can break some (atypical?) builds. (See the linked issue, which uses `FetchContent()`.

# Things to Scrutinize

- The option controls whether or not we install the mock libraries. It could control whether or not we define mock targets at all. Or whether we install headers.
  - naming follow up: it could be called `GOOGLE_CLOUD_CPP_INSTALL_MOCKS`.
- We are not exercising the option set to OFF in our CI.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13673)

### CODE REVIEW

**Comment by anthonyalayo** [comment]
I'm not a maintainer, but this is looking reasonable to me.

**Review by coryan — APPROVED** [review]

File: cmake/GoogleCloudCppCommon.cmake [hunk 1]
```diff
@@ -164,6 +164,17 @@ endfunction ()
 function (google_cloud_cpp_install_mocks library display_name)
     set(library_target "google_cloud_cpp_${library}")
     set(mocks_target "google_cloud_cpp_${library}_mocks")
+
+    # Always install the mock headers. These are harmless, in the sense that
+    # they will not cause build failures in environments without GoogleTest. The
```
[the remark this answers is not in the document]
Comment (anthonyalayo) [reply cmake/GoogleCloudCppCommon.cmake:169]: I assumed they were stubs from the comment, but if they're not, that would be problematic. I'm assuming other build issues creeped up when this wasn't done?

File: cmake/GoogleCloudCppLibrary.cmake [hunk 2]
```diff
@@ -349,7 +318,40 @@ function (google_cloud_cpp_add_gapic_library library display_name)
         DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/${library_target}"
         COMPONENT google_cloud_cpp_development)
 
-    google_cloud_cpp_install_mocks("${library}" "${display_name}")
+    if (GOOGLE_CLOUD_CPP_WITH_MOCKS)
+        # Create a header-only library for the mocks. We use a CMake `INTERFACE`
+        # library for these, a regular library would not work on macOS (where
+        # the library needs at least one .o file). Unfortunately INTERFACE
```
Comment (coryan) [inline cmake/GoogleCloudCppLibrary.cmake:324]: No action needed: I wonder if this is still true now that we use CMake >= 3.13 ..

File: cmake/GoogleCloudCppCommon.cmake [hunk 3]
```diff
@@ -164,6 +164,7 @@ endfunction ()
 function (google_cloud_cpp_install_mocks library display_name)
     set(library_target "google_cloud_cpp_${library}")
     set(mocks_target "google_cloud_cpp_${library}_mocks")
+
```
Comment (coryan) [inline cmake/GoogleCloudCppCommon.cmake:167]: Unrelated?

File: CMakeLists.txt [hunk 4]
```diff
@@ -165,6 +165,28 @@ This option is required to support package maintainers that split
...
+    [==[Build the google-cloud-cpp mocking libraries.
+
+google-cloud-cpp offers mocking libraries with mock classes, to facilitate unit
+testing of Cloud C++ clients. Consumers of this library that do not use the
+provided mocks to test code involving the Cloud C++ clients may wish to turn
+this flag off.]==]
+    ON
+    "NOT BUILD_TESTING"
+    ON)
+mark_as_advanced(GOOGLE_CLOUD_CPP_WITH_MOCKS)
```
Comment (coryan) [inline CMakeLists.txt:180]: TIL, you can use `NOT condition` in `cmake_dependent_option`.  Using `ON ... ON` looks weird.  I think you want something like this:


```suggestion
option(
    GOOGLE_CLOUD_CPP_WITH_MOCKS
    [==[Build the google-cloud-cpp mocking libraries.

google-cloud-cpp offers mocking libraries with mock classes, to facilitate unit
testing of Cloud C++ clients. Consumers of this library that do not use the
provided mocks to test code involving the Cloud C++ clients may wish to turn
this flag off.]==]
    $<IF:$<BOOL:${BUILD_TESTING}>:OFF:ON>)
mark_as_advanced(GOOGLE_CLOUD_CPP_WITH_MOCKS)
```

That is, the option always exists, but its default value is based on a generator expression.

**Review by coryan — APPROVED** [review]

File: cmake/GoogleCloudCppLibrary.cmake (hunk 2 above)
Comment (dbolduc) [reply cmake/GoogleCloudCppLibrary.cmake:324]: Ack, I don't know the answer and I don't have easy access to a mac.

**Comment by anthonyalayo** [comment]
Thanks for the merge!
